keps/sig-network/2086-service-internal-traffic-policy/README.md

```diff
@@ -123,51 +119,46 @@ type ServiceSpec struct {
 
 Proposed changes to kube-proxy:
-* when `trafficPolicy=Cluster`, default to existing behavior today.
-* when `trafficPolicy=Topology`, use topology hints from EndpointSlice API.
-* when `trafficPolicy=PreferLocal`, route to endpoints in EndpointSlice that matches the local node's topology (topology defined by `kubernetes.io/hostname`),
-fall back to "Cluster" behavior if there are no local endpoints.
-* when `trafficPolicy=Local`, route to endpoints in EndpointSlice that maches the local node's topology, drop traffic if none exist.
+* when `internalTrafficPolicy=Cluster`, default to existing behavior today.
+* when `internalTrafficPolicy=Local`, route to endpoints in EndpointSlice that maches the local node's topology, drop traffic if none exist.
 
 ### Test Plan
 
 Unit tests:
-* unit tests validating API strategy/validation for when `trafficPolicy` is set on Service.
-* unit tests exercising kube-proxy behavior when `trafficPolicy` is set to all possible values.
+* unit tests validating API strategy/validation for when `internalTrafficPolicy` is set on Service.
+* unit tests exercising kube-proxy behavior when `internalTrafficPolicy` is set to all possible values.
 
 E2E test:
-* e2e tests validating default behavior with kube-proxy did not change when `trafficPolicy` defaults to `Cluster`. Existing tests should cover this.
-* e2e tests validating that traffic is preferred to local endpoints when `trafficPolicy` is set to `PreferLocal`.
-* e2e tests validating that traffic is only sent to node-local endpoints when `trafficPolicy` is set to `Local`.
+* e2e tests validating default behavior with kube-proxy did not change when `internalTrafficPolicy` defaults to `Cluster`. Existing tests should cover this.
+* e2e tests validating that traffic is only sent to node-local endpoints when `internalTrafficPolicy` is set to `Local`.
 
 ### Graduation Criteria
 
 Alpha:
-* feature gate `ServiceTrafficPolicy` _must_ be enabled for apiserver to accept values for `spec.trafficPolicy`. Otherwise field is dropped.
-* kube-proxy handles traffic routing for 4 initial internal traffic policies `Cluster`, `Topology`, `PreferLocal` and `Local`.
+* feature gate `ServiceInternalTrafficPolicy` _must_ be enabled for apiserver to accept values for `spec.internalTrafficPolicy`. Otherwise field is dropped.
+* kube-proxy handles traffic routing for 2 initial internal traffic policies `Cluster`, and `Local`.
 * Unit tests as defined in "Test Plan" section above. E2E tests are nice to have but not required for Alpha.
 
+Beta:
+* integration tests exercising API behavior for `spec.internalTrafficPolicy` field of Service.
+* e2e tests exercising kube-proxy routing when `internalTrafficPolicy` is `Local`.
+* feature gate `ServiceInternalTrafficPolicy` is enabled by default.
+* consensus on how internalTrafficPolicy overlaps with topology-aware routing.
```

Comment:
TBH I lost track of where we landed on this specific topic, but we should probably reach consensus on the overlap prior to going to Beta.

Reply:
My current understanding is that we removed PreferLocal because it can now be handled with topology-aware routing (without endpointslice subsetting). What is our current thinking around how `internalTrafficPolicy: Local` and the new topology-aware API overlap? Are there plans for topology-aware routing to support something like `RequireNode`??

Reply:
Here's my understanding: *TrafficPolicy=Local always have precedence but can be used in conjunction with topology. Here's how I'm interpreting it anyway:

| ExternalTrafficPolicy | InternalTrafficPolicy | Topology | External Result | Internal Result |
| - | - | - | - | - |
| - | - | Auto | Topology | Topology |
| Local | - | Auto | Local | Topology |
| Local | Local | Auto | Local | Local |

Reply:
Thanks Rob! That makes sense to me.

Reply:
I like this table - it explains a lot. Can we put this into KEP?

Reply:
done!

Reply:
Thanks!